components/accounts/ios/accounts/FxAccountManager.swift

```diff
@@ -0,0 +1,361 @@
+
+    private init(raw: UInt64, serverConfig: FxAConfig, deviceConfig: DeviceConfig) {
+        self.raw = raw
+        self.serverConfig = serverConfig
+        self.deviceConfig = deviceConfig
+        deviceConstellation.parent = self
+    }
+
+    // TODO:
+    public convenience init(serverConfig: FxAConfig, deviceConfig: DeviceConfig) throws {
+        // TODO: in a helper!
+        var capabilities = MsgTypes_Capabilities()
+        deviceConfig.capabilities.forEach { capabilities.capability.append($0.toMessage()) }
+        let capabilitiesData = try! capabilities.serializedData()
+        let size = Int32(capabilitiesData.count)
+
+        let pointer = try queue.sync {
+            try capabilitiesData.withUnsafeBytes { (bytes: UnsafePointer<UInt8>) in
+                try FirefoxAccountError.unwrap { err in
+                    fxa_mgr_new(
+                        serverConfig.contentUrl,
+                        serverConfig.clientId,
+                        serverConfig.redirectUri,
+                        deviceConfig.name,
+                        Int32(deviceConfig.type.toMessage().rawValue),
+                        bytes,
+                        size,
+                        err
+                    )
+                }
+            }
+        }
+        self.init(raw: pointer, serverConfig: serverConfig, deviceConfig: deviceConfig)
+    }
+
+    deinit {
+        if self.raw != 0 {
+            queue.sync {
+                try! FirefoxAccountError.unwrap { err in
+                    // Is `try!` the right thing to do? We should only hit an error here
+                    // for panics and handle misuse, both inidicate bugs in our code
+                    // (the first in the rust code, the 2nd in this swift wrapper).
+                    fxa_mgr_free(self.raw, err)
+                }
+            }
+        }
+    }
+
+    public var accountState: AccountState {
+        return queue.sync {
+            do {
+                let stateBuffer = try FirefoxAccountError.unwrap { err in
+                    fxa_mgr_account_state(self.raw, err)
+                }
+                let msg = try! MsgTypes_AccountState(serializedData: Data(rustBuffer: stateBuffer))
+                fxa_mgr_bytebuffer_free(stateBuffer)
+                return AccountState(msg: msg)
+            } catch {
+                os_log("Error while calling fxa_mgr_account_state")
+                return .Start
+            }
+        }
+    }
+
+    public func initialize() {
+        queue.async { // The do-catch inside queue.async pattern may be factorized?
+            do {
+                let jsonState = self.accountStorage.read()
+                try FirefoxAccountError.unwrap { err in
+                    fxa_mgr_init(self.raw, jsonState, err)
+                }
+                self.persistAccount()
+                switch self.accountState {
+                case .Authenticated: self.postAuthenticated()
```

Comment:
IIUC, we end up firing an `onAuthenticated` event here not because something happened that caused us to become authenticated, but because we've entered the "authenticated" state on startup. I can see callers potentially being confused by this and e.g. showing some sort of success notification as though the user had just logged in.

I wonder, is the abstraction you're actually after here an "on state change" event with the resulting state as argument? Or are there cases where we enter one of these states without firing a notification?